File: App/SpectralPeakText.swift
import Foundation
import TurntableCore

/// 譜峰判讀的文案。
///
/// **文字放在 App 層，不在核心。** 核心回傳的是分類（`SpectralPeak.Kind`），
/// 由這裡翻成給人看的句子 —— 這樣核心保持純演算法、不必背本地化資源，
/// 而文案也才能跟著系統語言走。
extension SpectralPeak {
    var interpretation: String {
        Self.text(for: kind, order: orderOfRotation)
    }

    static func text(for kind: Kind, order: Double) -> String {
        switch kind {
        case .eccentricity:
            return String(localized: "每圈一次 —— 偏心（盤面、主軸或皮帶接觸面沒對正）")
        case .ovality:
            // **1× 大的時候這個判讀不可信。** 非正弦的每圈一次擾動本來就會生出 2×、3×。
            // 實測把手機轉 180° 讓 1× 掉 74% 之後，2× 也跟著掉 59% —— 而純粹的
            // 180° 旋轉不該改變 2×（相位移 360°），所以那是耦合的諧波不是獨立缺陷。
            return String(localized: "每圈兩次 —— 盤面橢圓或主軸兩點磨損。但如果上面的每圈一次很大，這根多半只是它的諧波，先把偏心降下來再看。")
        case .harmonic(let n):
            return String(localized: "轉盤 \(n)× 諧波")
        case .slowerThanRotation:
            return String(localized: "比一圈還慢 —— 皮帶循環或長週期漂移")
        case .driveChain:
            let times = String(format: "%.1f", order)
            return String(localized: "非諧波（轉盤的 \(times) 倍）—— 傳動鏈上的零件，馬達或皮帶輪的候選")
        }
    }
}

extension StoredPeak {
    /// 存檔用的簡短版本。歷史列表空間比較窄。
    var interpretation: String {
        let n = Int(orderOfRotation.rounded())
        if isHarmonic {
            switch n {
            case 1:  return String(localized: "每圈一次 —— 偏心")
            case 2:  return String(localized: "每圈兩次 —— 盤面橢圓")
            default: return String(localized: "轉盤 \(n)× 諧波")
            }
        }
        if orderOfRotation < 1 { return String(localized: "比一圈還慢 —— 皮帶循環") }
        let times = String(format: "%.1f", orderOfRotation)
        return String(localized: "非諧波（\(times) 倍）—— 傳動鏈")
    }
}

File: android/app/src/main/java/com/roger/turntablerpm/ui/SpectralPeakText.kt
package com.roger.turntablerpm.ui

import androidx.compose.runtime.Composable
import androidx.compose.ui.res.stringResource
import com.roger.turntablerpm.R
import com.roger.turntablerpm.core.SpectralPeak

/**
 * 譜峰判讀的文案。
 *
 * **核心回傳的是分類（`SpectralPeak.Kind`），文字在這裡組。** 文案屬於 UI 層，
 * 而且要能本地化 —— 把中文寫死在演算法核心裡，等於逼那個純 Kotlin 模組去背資源檔。
 * iOS 端是同樣的分工（`SpectralPeakText.swift`）。
 *
 * 目前先寫死中文。要多語系時搬進 `strings.xml` 即可 —— Android 的資源查找是顯式的，
 * 不會像 iOS 的字串抽取那樣靜默失敗（見 CLAUDE.md 坑 29）。
 */
@Composable
fun interpretation(peak: SpectralPeak): String = when (val k = peak.kind) {
    SpectralPeak.Kind.Eccentricity ->
        stringResource(R.string.peak_eccentricity)
    // **1× 大的時候這個判讀不可信。** 非正弦的每圈一次擾動本來就會生出 2×、3×。
    // 實測把手機轉 180° 讓 1× 掉 74% 之後，2× 也跟著掉 59% —— 而純粹的 180° 旋轉
    // 不該改變 2×（相位移 360°），所以那是耦合的諧波不是獨立缺陷。
    SpectralPeak.Kind.Ovality ->
        stringResource(R.string.peak_ovality)
    is SpectralPeak.Kind.Harmonic ->
        stringResource(R.string.peak_harmonic, k.order)
    SpectralPeak.Kind.SlowerThanRotation ->
        stringResource(R.string.peak_slower_than_rotation)
    SpectralPeak.Kind.DriveChain ->
        stringResource(R.string.peak_drive_chain, peak.orderOfRotation)
}

/** 整數倍代表跟著盤面轉的東西，非整數倍代表傳動鏈 —— 用顏色區分。 */
fun isHarmonic(peak: SpectralPeak): Boolean = peak.isRotationHarmonic

/**
 * 存檔用譜峰的簡短判讀。
 *
 * **跟核心的 `SpectralPeak` 分開，因為存檔格式跟核心型別是分開演進的**
 * （見 `StoredPeak` 的註解）。歷史頁的空間也比較窄，所以句子短一些。
 */
@Composable
fun storedInterpretation(peak: com.roger.turntablerpm.history.StoredPeak): String {
    val n = Math.round(peak.orderOfRotation).toInt()
    if (peak.isHarmonic) {
        return when (n) {
            1 -> stringResource(R.string.stored_peak_eccentricity)
            2 -> stringResource(R.string.stored_peak_ovality)
            else -> stringResource(R.string.stored_peak_harmonic, n)
        }
    }
    if (peak.orderOfRotation < 1) return stringResource(R.string.stored_peak_slower)
    return stringResource(R.string.stored_peak_drive_chain, peak.orderOfRotation)
}
